### PR TITLE
feat(menu-bar): Enhanced visual feedback for "Mark me as done"

### PR DESCRIPTION
**Pull request by liza-kl**
## Description
- Implemented the enhanced feedback behaviour for the "Mark me as Done" tooltip, as stated in [#4269](https://github.com/inovex/scrumlr.io/issues/4269).

- The cases are handled as follows (or at least this was the intention):
• **Voting is active / No timer** → Showing the open tooltip after 2 seconds and leaving it open until the user clicks. Clicking logs the toggle interaction. If a user decides to unmark them, the feedback shall not be shown. 
• **Voting is active / timer is active / votes are used up **before** the timer ends** → Showing the tooltip after 2 seconds and leaving it open a) until the user clicks or b) until 30 seconds after the timer expires and the user is set to ready automatically 

• **Voting not active / Timer is active** → After timer expires and the user is not ready, showing the tooltip after 2 seconds until 30 seconds after expiration and then setting the user automatically to done. 

• **Voting is active / timer has expired / votes are not used up**: same behaviour as _Voting is not active, but timer is active_ 
• Voting not active, timer active, user toggles their state twice within 30 seconds after expiration -> don't show the open tooltip for the second time.  

## Changelog

- Add hooks that implement logic in `MenuBars.tsx`
- Add tests for implemented logic 
## Checklist

- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] The light- and dark-theme are both supported and tested
- [x] The design was implemented and is responsive for all devices and screen sizes
- [x] The application was tested in the most commonly used browsers (e.g. Chrome, Firefox, Safari)

### CODE REVIEW

**Review by Schwehn42 — COMMENTED** [review]
I'm admittedly kinda confused by the behaviour:

- when a timer is active, all participants get set to ready after 30secs. The 30s don't start after the timer has expired but after the timer was started, which can't be right, can it?
- that is also the case when a voting is active, and it doesn't matter whether one actually votes
- the timer itself doesn't seem to disappear after 30s after expiration

**Comment by brandstetterm** [comment]
> b) until 30 seconds after the timer expires and the user is set to ready automatically

Nice work! Two questions: Why do participants get set to ready 30s after the timer has ended automatically? What makes you think that the competitor is in fact ready?

File: src/utils/hooks/useTimerLeft.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import {useEffect, useState} from "react";
+import {Timer as TimerUtils} from "../timer";
+
+/**
+ * A hook which calculates the time left based on a provided end Date.
+ * Intended for: https://github.com/inovex/scrumlr.io/issues/4217
+ * @param timerEnd T
```
Comment (brandstetterm) [inline src/utils/hooks/useTimerLeft.ts:7]: The explanation of the parameter seems to be missing

File: src/utils/hooks/useTimerLeft.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      } else {
+        timer = setTimeout(updateTimer, 250);
+        setTimerExpired(false);
+      }
+    };
+    updateTimer();
+    return () => clearTimeout(timer);
+  }, [timerEnd]);
+
+  return {timerExpired};
```
Comment (brandstetterm) [inline src/utils/hooks/useTimerLeft.ts:31]: Is it possible to only return the boolean value of `timerExpired` or does it have to be wrapped within an object?

File: src/components/TooltipButton/TooltipButton.tsx [hunk 3]
```diff
@@ -22,6 +23,7 @@ export const TooltipButton = (props: TooltipButtonProps) => {
       className={classNames("tooltip-button", `tooltip-button--${props.direction ?? "right"}`, {"tooltip-button--active": props.active}, props.className)}
       onClick={() => props.onClick()}
       aria-label={props.label}
+      data-testid={props.dataTestid}
```
Comment (brandstetterm) [inline src/components/TooltipButton/TooltipButton.tsx:26]: Using data test ids should be avoided if possible since it doesn't resemble how our application is used

File: src/components/MenuBars/__tests__/MenuBars.test.tsx [hunk 4]
```diff
@@ -47,3 +62,54 @@ describe("MenuBars", () => {
...
+    jest.useFakeTimers();
+    (useTimer as jest.Mock).mockReturnValue({timerExpired: true});
+    const store = getTestStore({
+      participants: {
+        self: getTestParticipant({ready: true}),
+        others: [],
+      },
+    });
+    const {getByTestId} = render(createMenuBars(store));
+    const button = getByTestId("mark-as-done-button");
```
Comment (brandstetterm) [inline src/components/MenuBars/__tests__/MenuBars.test.tsx:106]: Using test ids should be avoided if possible, since it doesn't resemble the real world usage of our application. It always better to test from a user perspective => Using text or labels to get elements.

**Comment by liza-kl** [comment]
> > b) until 30 seconds after the timer expires and the user is set to ready automatically
> 
> Nice work! Two questions: Why do participants get set to ready 30s after the timer has ended automatically? What makes you confident that the participants are in fact ready?

This was a technical requirement (No. 3) stated in #4269. However, I see that I haven't implemented the "closes the timer automatically". Shall we remove the 30s rule then in general?

**Comment by brandstetterm** [comment]
> This was a technical requirement (No. 3) stated in #4269. However, I see that I haven't implemented the "closes the timer automatically". Shall we remove the 30s rule then, in general?

Thanks for the clarification! To be honest, I'm not really keen on automatically marking participants as ready, as we can never be sure that they are actually ready. However, I also don't want to simply change requirements for features that @SelinaBuff has defined. I think I'm going to start a discussion in our Slack channel to find out what the others think.

File: src/components/MenuBars/__tests__/MenuBars.test.tsx (hunk 4 above)
Comment (liza-kl) [reply src/components/MenuBars/__tests__/MenuBars.test.tsx]: Replaced by using text now :)

File: src/components/TooltipButton/TooltipButton.tsx (hunk 3 above)
Comment (liza-kl) [reply src/components/TooltipButton/TooltipButton.tsx]: Removed it

**Comment by liza-kl** [comment]
> > This was a technical requirement (No. 3) stated in #4269. However, I see that I haven't implemented the "closes the timer automatically". Shall we remove the 30s rule then, in general?
> 
> Thanks for the clarification! To be honest, I'm not really keen on automatically marking participants as ready, as we can never be sure that they are actually ready. However, I also don't want to simply change requirements for features that @SelinaBuff has defined. I think I'm going to start a discussion in our Slack channel to find out what the others think.

That's a good idea. Maybe we should also discuss the edge case @Schwehn42 pointed out :)

**Review by Schwehn42 — COMMENTED** [review]
Still some issues with it, especially when adding a new timer or resetting it after it ran out. Didn't test with the votes yet, just the timer.
Hard reset means concluding the current timer before starting a new one, while soft reset means just initializing a timer, even though the old one is still active (despite being at 0s).

| Use Case         | Shows Banner |
|------------------|--------------|
| initial timer     | ✅            |
| mid soft reset   | ✅            |
| mid add +1       | ✅            |
| post soft reset  | ⛔            |
| post add +1      | ⛔            |
| post hard reset  | ⛔            |

**Comment by liza-kl** [comment]
- New "simplified" logic (was discussed and approved)
  - Everytime a condition for a ready state is fulfilled (e.g. timer expired, or votes are used up) show the extended tooltip after 2 seconds for another 28 seconds.

**Review by laila-rin — APPROVED** [review]
Looks good!
